File: src/go/sliding_window/567.permutation-in-string.go
package main

func checkInclusion(s1 string, s2 string) bool {
	if len(s1) > len(s2) {
		return false
	}
	count := [26]int{}
	for i := 0; i < len(s1); i++ {
		count[s1[i]-'a']++
		count[s2[i]-'a']--
	}
	if allZero(count[:]) {
		return true
	}
	for i := len(s1); i < len(s2); i++ {
		count[s2[i]-'a']--
		count[s2[i-len(s1)]-'a']++
		if allZero(count[:]) {
			return true
		}
	}
	return false
}

File: src/java/sliding_window/567.permutation-in-string.java
class P567 {
    // @lc app=leetcode id=567 lang=java
    // @lc code=start
    public class Solution {
        public boolean checkInclusion(String s1, String s2) {
            int n1 = s1.length(), n2 = s2.length();
            if (n1 > n2) return false;
            int[] need = new int[26], have = new int[26];
            for (int i=0;i<n1;i++) need[s1.charAt(i)-'a']++;
            for (int i=0;i<n2;i++){
                have[s2.charAt(i)-'a']++;
                if (i>=n1) have[s2.charAt(i-n1)-'a']--;
                if (i+1>=n1 && same(have, need)) return true;
            }
            return false;
        }
        private boolean same(int[] a, int[] b){ for(int i=0;i<26;i++) if (a[i]!=b[i]) return false; return true; }
    }
    // @lc code=end
}
